upgrade/upgrade-scripts/execute

```diff
@@ -63,8 +68,23 @@ EOF
 
 apt_get update || die "failed to update apt sources"
 
+#
+# First make sure delphix-platform-<platform> is installed, matching the
+# installation order used in the appliance-build. This has to be done
+# before installing delphix-entire, because delphix-entire depends on
+# delphix-virtualization which in turn depends on the virtual package
+# delphix-platform. If some version of delphix-platform is already
+# installed, that dependency is already satisfied. If it is not, apt
+# will choose one of the packages that satisfies the dependency, and it
+# might choose the wrong one. When this happens apt will end up trying
+# to install two packages which provide delphix-platform, which will
+# fail because the two packages conflict with each other.
+#
+apt_get install -y "delphix-platform-$platform" ||
```

Comment:
@jgallag88 I know we've talked about this and I agreed on the approach, I'd still really like to understand more why is apt getting confused by this. 
Since `delphix-entire-$platform` should depend on `delphix-platform-$platform` and `delphix-platform-$platform` already provides `delphix-platform`, I'm really surprised that apt doesn't take it into consideration when looking for the dependencies of `delphix-virtualization`.
Let me know if you have a system in that state, I'd really love to take a look at it and understand better what's going on.

Reply:
> If we leave this out, why doesn't installing delphix-entire-$platform install delphix-platform-$platform? doesn't delphix-entire-$platform depend on delphix-platform-$platform?

So it does install delphix-platform-$platform, but it installs delphix-platform-$someOtherPlatform too. Say we are upgrading a vm on AWS. I suspect what happens is that apt first sees the `delphix-virtualization` dependency on `delphix-platform`. It also sees that there are mulitple versions of `delphix-platform` available, none of which are yet installed. In this scenario, apt picks an arbitrary package that provides the virtual package to install, say `delphix-platform-kvm`. It later sees that it needs to install the concrete package `delphix-platform-aws` because that is a dependency of `delphix-entire-aws`. It isn't smart enough to go back and realize that it no longer needs `delphix-platform-kvm`, so it tries to install both, and they conflict with each other.

It's possible that `apt_get install -y "delphix-platform-$platform delphix-entire-$platform` would work, but I don't know how we could be sure whether it's guaranteed to work or is just working accidentally unless we understood all of the details about how apt resolves transitive dependencies where virtual packages are involved (for which I haven't found any documentation).

Another way to resolve this would be to remove the explicit dependency on `delphix-platform` from `delphix-virtualization`. What I like about the current strategy though is that it makes the order in which the packages are installed more similar to the order used in live-build in appliance-build. I think this is good from a consistency perspective, since in general installing the same packages in different orders can result in different end-states. However, there could be other ways to make appliance build more similar to a not-in-place upgrade, maybe by getting rid of live-build, as you mentioned elsewhere.

@pzakha I can get a VM set up shortly.

Reply:
I'm not 100% against installing `delphix-platform-$platform` prior to installing `delphix-entire-$platform`, but it does set a precedence that I'd like to avoid if we can.. since, this means if we ever want to create some other platform specific package(s), upgrade will need explicit knowledge about them. e.g. if we create some new package(s) `foo-aws`, `foo-esx`, etc. upgrade will have to know how to install the correct one, right?

Ultimately, I'd like to see if we can understand this behavior prior to coming to a decision.

> What I like about the current strategy though is that it makes the order in which the packages are installed more similar to the order used in live-build in appliance-build.

Yea, I agree with this in concept, but I'd argue the build is worse about this than upgrade. I'd rather see us revamp the build, to avoid the step-by-step implementation that it has now. If we don't change how the build works, though, I think this is a good argument for taking this approach.

Reply:
Based on our offline findings and discussion, I've expanded the comment here substantially to explain exactly what goes wrong if we install delphix-entire-<platform> directly.